lib/reek/source/source_locator.rb

```diff
@@ -16,22 +18,37 @@ def initialize(paths)
       #
       # Returns a list of Source::SourceCode.
       def sources
-        find_sources.map { |pathname| Source::SourceCode.from File.new(pathname) }
+        source_paths.map { |pathname| Source::SourceCode.from File.new(pathname) }
       end
 
       private
 
-      def find_sources(paths = @paths)
-        paths.map do |path|
-          pathname = Pathname.new(path)
-          if pathname.directory?
-            find_sources(Dir["#{pathname}/**/*.rb"])
-          else
-            next pathname if pathname.file?
-            $stderr.puts "Error: No such file - #{pathname}"
-            nil
+      def source_paths
+        relevant_paths = []
+        @paths.map do |given_path|
+          print_no_such_file_error(given_path) && next unless path_exists?(given_path)
+          Find.find(given_path) do |path|
```

Comment:
Somebody using `reek` in one thread and _anything else_ that uses `Find.prune` in another thread. My very quick-and-dirty proof of concept:

```Ruby
def prune
  throw :prune
end

a = Thread.new do
  puts 'a'
  prune
end

b = Thread.new do
  puts 'b'
  catch(:prune) { puts 'caught' }
end

a.join
b.join
```

– if you run this enough times you’ll see `caught` being printed.

That said, I’m not sure whether `reek` is anywhere to being thread-safe everywhere else; I don’t want this to block this PR as long as we track the knowledge somewhere and come back to this issue eventually. :)